builder/amazon/common/ssm_driver.go

```diff
@@ -0,0 +1,117 @@
+package common
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"os/exec"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/service/ssm"
+)
+
+const sessionManagerPluginName string = "session-manager-plugin"
+
+//sessionCommand is the AWS-SDK equivalent to the command you would specify to `aws ssm ...`
+const sessionCommand string = "StartSession"
+
+type SSMDriver struct {
+	Region          string
+	ProfileName     string
+	Session         *ssm.StartSessionOutput
+	SessionParams   ssm.StartSessionInput
+	SessionEndpoint string
+	// Provided for testing purposes; if not specified it defaults to sessionManagerPluginName
+	PluginName string
+}
+
+// StartSession starts an interactive Systems Manager session with a remote instance via the AWS session-manager-plugin
+func (sd *SSMDriver) StartSession(ctx context.Context) error {
+	if sd.PluginName == "" {
+		sd.PluginName = sessionManagerPluginName
+	}
+
+	args, err := sd.Args()
+	if err != nil {
+		err = fmt.Errorf("error encountered validating session details: %s", err)
+		return err
+	}
+
+	cmd := exec.CommandContext(ctx, sd.PluginName, args...)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return err
+	}
+
+	// Aggregate all output into one reader
+	combinedOut := io.MultiReader(stdout, stderr)
+
+	if err := cmd.Start(); err != nil {
+		err = fmt.Errorf("error encountered when calling %s: %s\n", sd.PluginName, err)
+		return err
+	}
+
+	output := bufio.NewScanner(combinedOut)
+	successLogLine := fmt.Sprintf("opened for sessionId %s", *sd.Session.SessionId)
+	for output.Scan() {
+		if output.Err() != nil && output.Err() != io.EOF {
+			break
+		}
+
+		out := output.Text()
+		if out != "" {
+			if strings.Contains(out, "panic") {
+				line := fmt.Sprintf("[%s stderr] %s\n", sd.PluginName, out)
+				log.Print(line)
+				return fmt.Errorf("exited with a non-zero status")
+			}
+
+			line := fmt.Sprintf("[%s] %s\n", sd.PluginName, out)
+			log.Print(line)
+
+			if strings.Contains(line, successLogLine) {
+				return nil
+			}
+		}
+	}
```

Comment:
I think this whole loop should be in a goroutine because otherwise `StartSession` will never return 

Also it's interesting to note: 
> StdoutPipe returns a pipe that will be connected to the remote command's standard output when the command starts. There is a fixed amount of buffering that is shared between stdout and stderr streams. If the StdoutPipe reader is not serviced fast enough it *may eventually cause the remote command to block*.

So that means a resources starvation could cause the command to block while printing things.

Reply:
> I think this whole loop should be in a goroutine because otherwise StartSession will never return

Good catch, I realized that if we don't log anything we a blocked and thought of adding a timer that would break if it couldn't be determined if the tunnel was open. But "So that means a resources starvation could cause the command to block while printing things" makes me question things a bit.

Reply:
Adrien I think that's not true -- the loop gets broken out of so the session returns. This isn't a long-running log streaming thing; once the session is up, we break out of the loop because it won't log anymore, and we don't care. The end is always either an error causing the loop to break, or a success message. And we don't want to return until one of those cases is fulfilled. 

Resource starvation also isn't an issue -- this tunnel's output is super sparse.